From 2391a2633cc40d5826e74fe8f146b1861bd80c5c Mon Sep 17 00:00:00 2001
From: David Henningsson <david.henningsson@canonical.com>
Date: Tue, 27 Aug 2013 07:04:37 +0200
Subject: [PATCH 1/2] Add UCM configuration for Nexus 4 (Mako)

The VoiceCall profile is mostly empty, because in my implementation,
this is handled by talking to the audio HAL.
---
 configure.in                                       |    1 +
 src/conf/ucm/Makefile.am                           |    2 +-
 src/conf/ucm/apq8064-tabla-snd-card/HiFi           |  259 ++++++++++++++++++++
 src/conf/ucm/apq8064-tabla-snd-card/Makefile.am    |    4 +
 src/conf/ucm/apq8064-tabla-snd-card/VoiceCall      |  121 +++++++++
 .../apq8064-tabla-snd-card.conf                    |    8 +
 6 files changed, 394 insertions(+), 1 deletion(-)
 create mode 100644 src/conf/ucm/apq8064-tabla-snd-card/HiFi
 create mode 100644 src/conf/ucm/apq8064-tabla-snd-card/Makefile.am
 create mode 100644 src/conf/ucm/apq8064-tabla-snd-card/VoiceCall
 create mode 100644 src/conf/ucm/apq8064-tabla-snd-card/apq8064-tabla-snd-card.conf

diff --git a/configure.in b/configure.in
index 00fff2b..25fdc81 100644
--- a/configure.in
+++ b/configure.in
@@ -669,6 +669,7 @@ AC_OUTPUT(Makefile doc/Makefile doc/pictures/Makefile doc/doxygen.cfg \
 	  src/conf/ucm/PandaBoardES/Makefile \
 	  src/conf/ucm/SDP4430/Makefile \
 	  src/conf/ucm/tegraalc5632/Makefile \
+	  src/conf/ucm/apq8064-tabla-snd-card/Makefile \
 	  modules/Makefile modules/mixer/Makefile modules/mixer/simple/Makefile \
 	  alsalisp/Makefile aserver/Makefile \
 	  test/Makefile test/lsb/Makefile \
diff --git a/src/conf/ucm/Makefile.am b/src/conf/ucm/Makefile.am
index e089dde..adb3c87 100644
--- a/src/conf/ucm/Makefile.am
+++ b/src/conf/ucm/Makefile.am
@@ -1 +1 @@
-SUBDIRS=DAISY-I2S PandaBoard PandaBoardES SDP4430 tegraalc5632
+SUBDIRS=DAISY-I2S PandaBoard PandaBoardES SDP4430 tegraalc5632 apq8064-tabla-snd-card
diff --git a/src/conf/ucm/apq8064-tabla-snd-card/HiFi b/src/conf/ucm/apq8064-tabla-snd-card/HiFi
new file mode 100644
index 0000000..a5db5a9
--- /dev/null
+++ b/src/conf/ucm/apq8064-tabla-snd-card/HiFi
@@ -0,0 +1,259 @@
+SectionVerb {
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 0"
+	]
+
+	Value {
+		TQ "HiFi"
+		
+		CapturePCM "hw:apq8064tablasnd,0"
+		PlaybackPCM "hw:apq8064tablasnd,0"
+	}
+}
+
+SectionDevice."Earpiece" {
+	#Name "Earpiece"
+	Comment "Earpiece speaker"
+
+	# FIXME
+	ConflictingDevice [
+		"Headphones"
+		"Speaker"
+	]
+
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# We need to disable this while setting up the route, else the
+		# driver will hang if the stream is open.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='SLIM_0_RX Channels' One"
+		cset "name='RX1 MIX1 INP1' RX1"
+		cset "name='RX1 MIX1 INP2' RX1"
+		cset "name='RX1 Digital Volume' 68"
+		cset "name='EAR PA Gain' POS_6_DB"
+		cset "name='DAC1 Switch' 1"
+		cset "name='SLIM TX3 MUX' RMIX1"
+
+		# Route has been set up, re-enable the mixer.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='RX1 MIX1 INP1' ZERO"
+		cset "name='RX1 MIX1 INP2' ZERO"
+		cset "name='RX1 Digital Volume' 68"
+		cset "name='DAC1 Switch' 0"
+		cset "name='SLIM TX3 MUX' ZERO"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Speaker" {
+	#Name "Speaker"
+	Comment "Main speaker"
+
+	# FIXME
+	ConflictingDevice [
+		"Earpiece"
+		"Headphones"
+	]
+
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# We need to disable this while setting up the route, else the
+		# driver will hang if the stream is open.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='SLIM_0_RX Channels' One"
+		cset "name='RX3 MIX1 INP1' RX1"
+		cset "name='RX3 MIX1 INP2' RX6"
+		cset "name='RX4 DSM MUX' DSM_INV"
+		cset "name='RX3 Digital Volume' 68"
+		cset "name='LINEOUT1 Volume' 100"
+		cset "name='LINEOUT3 Volume' 100"
+		cset "name='SLIM TX3 MUX' RMIX3"
+
+		# Route has been set up, re-enable the mixer.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='RX3 MIX1 INP1' ZERO"
+		cset "name='RX3 MIX1 INP2' ZERO"
+		cset "name='RX4 DSM MUX' CIC_OUT"
+		cset "name='RX3 Digital Volume' 68"
+		cset "name='LINEOUT1 Volume' 0"
+		cset "name='LINEOUT3 Volume' 0"
+		cset "name='SLIM TX3 MUX' ZERO"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Headphones" {
+	#Name "Headphones"
+	Comment "Headphones"
+
+	# FIXME
+	ConflictingDevice [
+		"Earpiece"
+		"Speaker"
+	]
+
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# We need to disable this while setting up the route, else the
+		# driver will hang if the stream is open.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='SLIM_0_RX Channels' Two"
+		cset "name='RX1 MIX1 INP1' RX1"
+		cset "name='RX1 MIX1 INP2' RX6"
+		cset "name='RX2 MIX1 INP1' RX2"
+		cset "name='RX2 MIX1 INP2' RX7"
+		cset "name='HPHL DAC Switch' 1"
+		#FIXME ENOEXIST: cset "name='HPHR DAC Switch' 1"
+		cset "name='RX1 Digital Volume' 68"
+		cset "name='RX2 Digital Volume' 68"
+		cset "name='HPHL Volume' 50"
+		cset "name='HPHR Volume' 50"
+		cset "name='SLIM TX3 MUX' RMIX1"
+		cset "name='SLIM TX5 MUX' RMIX2"
+
+		# Route has been set up, re-enable the mixer.
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+
+		cset "name='RX1 MIX1 INP1' ZERO"
+		cset "name='RX1 MIX1 INP2' ZERO"
+		cset "name='RX2 MIX1 INP1' ZERO"
+		cset "name='RX2 MIX1 INP2' ZERO"
+		cset "name='HPHL DAC Switch' 0"
+		#FIXME ENOEXIST: cset "name='HPHR DAC Switch' 0"
+		cset "name='RX1 Digital Volume' 68"
+		cset "name='RX2 Digital Volume' 68"
+		cset "name='HPHL Volume' 0"
+		cset "name='HPHR Volume' 0"
+		cset "name='SLIM TX3 MUX' ZERO"
+		cset "name='SLIM TX5 MUX' ZERO"
+
+		# See comment in EnableSequence
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Headset" {
+	#Name "Headset"
+	Comment "Headset Microphone"
+
+	ConflictingDevice [
+		"Handset"
+	]
+
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 0"
+		cset "name='SLIM_0_TX Channels' One"
+		cset "name='SLIM TX7 MUX' DEC8"
+		cset "name='DEC8 MUX' ADC2"
+		cset "name='ADC2 Volume' 3"
+		cset "name='DEC8 Volume' 84"
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 0"
+		cset "name='SLIM TX7 MUX' ZERO"
+		cset "name='DEC8 MUX' ZERO"
+		cset "name='ADC2 Volume' 0"
+	]
+
+	Value {
+		CaptureChannels "1"
+	}
+}
+
+SectionDevice."Handset" {
+	#Name "Handset"
+	Comment "Handset Microphone"
+
+	ConflictingDevice [
+		"Headset"
+	]
+
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 0"
+		cset "name='SLIM_0_TX Channels' One"
+		cset "name='SLIM TX7 MUX' DEC7"
+		cset "name='DEC7 MUX' ADC1"
+		cset "name='DEC7 Volume' 80"
+		cset "name='ADC1 Volume' 100"
+		cset "name='MICBIAS1 CAPLESS Switch' 0"
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+
+		cset "name='MultiMedia1 Mixer SLIM_0_TX' 0"
+		cset "name='SLIM TX7 MUX' ZERO"
+		cset "name='DEC7 MUX' ZERO"
+		cset "name='DEC7 Volume' 0"
+		cset "name='ADC1 Volume' 0"
+		cset "name='MICBIAS1 CAPLESS Switch' 1"
+	]
+
+	Value {
+		CaptureChannels "1"
+	}
+}
diff --git a/src/conf/ucm/apq8064-tabla-snd-card/Makefile.am b/src/conf/ucm/apq8064-tabla-snd-card/Makefile.am
new file mode 100644
index 0000000..71309db
--- /dev/null
+++ b/src/conf/ucm/apq8064-tabla-snd-card/Makefile.am
@@ -0,0 +1,4 @@
+alsaconfigdir = @ALSA_CONFIG_DIR@
+ucmdir = $(alsaconfigdir)/ucm/apq8064-tabla-snd-card
+ucm_DATA = apq8064-tabla-snd-card.conf HiFi VoiceCall
+EXTRA_DIST = $(ucm_DATA)
diff --git a/src/conf/ucm/apq8064-tabla-snd-card/VoiceCall b/src/conf/ucm/apq8064-tabla-snd-card/VoiceCall
new file mode 100644
index 0000000..dc2007d
--- /dev/null
+++ b/src/conf/ucm/apq8064-tabla-snd-card/VoiceCall
@@ -0,0 +1,121 @@
+# There is no mixer setup here - it is handled by PulseAudio's audio HAL module.
+
+SectionVerb {
+	EnableSequence [
+		cdev "hw:apq8064tablasnd"
+		cset "name='MultiMedia1 Mixer SLIM_4_TX' 1"
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 1"
+	]
+
+	DisableSequence [
+		cdev "hw:apq8064tablasnd"
+		cset "name='MultiMedia1 Mixer SLIM_4_TX' 0"
+		cset "name='SLIMBUS_0_RX Audio Mixer MultiMedia1' 0"
+	]
+
+	Value {
+		CapturePCM "hw:apq8064tablasnd,0"
+		PlaybackPCM "hw:apq8064tablasnd,0"
+	}
+}
+
+SectionDevice."Earpiece" {
+	#Name "Earpiece"
+	Comment "Earpiece speaker"
+
+	# FIXME
+	ConflictingDevice [
+		"Headphones"
+		"Speaker"
+	]
+
+	EnableSequence [
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Speaker" {
+	#Name "Speaker"
+	Comment "Main speaker"
+
+	# FIXME
+	ConflictingDevice [
+		"Earpiece"
+		"Headphones"
+	]
+
+	EnableSequence [
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Headphones" {
+	#Name "Headphones"
+	Comment "Headphones"
+
+	# FIXME
+	ConflictingDevice [
+		"Earpiece"
+		"Speaker"
+	]
+
+	EnableSequence [
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+		PlaybackChannels "2"
+	}
+}
+
+SectionDevice."Headset" {
+	#Name "Headset"
+	Comment "Headset Microphone"
+
+	ConflictingDevice [
+		"Handset"
+	]
+
+	EnableSequence [
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+		CaptureChannels "1"
+	}
+}
+
+SectionDevice."Handset" {
+	#Name "Handset"
+	Comment "Handset Microphone"
+
+	ConflictingDevice [
+		"Headset"
+	]
+
+	EnableSequence [
+	]
+
+	DisableSequence [
+	]
+
+	Value {
+		CaptureChannels "1"
+	}
+}
diff --git a/src/conf/ucm/apq8064-tabla-snd-card/apq8064-tabla-snd-card.conf b/src/conf/ucm/apq8064-tabla-snd-card/apq8064-tabla-snd-card.conf
new file mode 100644
index 0000000..a4a7c1c
--- /dev/null
+++ b/src/conf/ucm/apq8064-tabla-snd-card/apq8064-tabla-snd-card.conf
@@ -0,0 +1,8 @@
+SectionUseCase."HiFi" {
+	File "HiFi"
+	Comment "Play HiFi quality Music."
+}
+SectionUseCase."Voice Call" {
+	File "VoiceCall"
+	Comment "Make a phone call"
+}
-- 
1.7.9.5

